system/jlib/jmetrics.hpp

```diff
@@ -237,6 +236,101 @@ class CustomMetric : public MetricBase
 };
 
 
+template <typename T>
```

Comment:
I don't think the template parameter is a good idea for a few reasons
* It complicates all the types, and I'm not sure it adds much
* The sinks need to be able to get at the results - they need a virtual function in IMetric that they can call to get them, and that is not possible if the type is a template.